crates/ra_hir/src/ty/infer.rs

```diff
@@ -54,13 +54,33 @@ pub fn infer(db: &impl HirDatabase, func: Function) -> Arc<InferenceResult> {
     Arc::new(ctx.resolve_all())
 }
 
+#[derive(Debug, Copy, Clone)]
+enum ExprOrPatId {
+    Expr(ExprId),
+    Pat(PatId),
+}
+
+impl From<ExprId> for ExprOrPatId {
+    fn from(id: ExprId) -> Self {
+        ExprOrPatId::Expr(id)
+    }
+}
+
+impl From<PatId> for ExprOrPatId {
+    fn from(id: PatId) -> Self {
+        ExprOrPatId::Pat(id)
+    }
+}
+
 /// The result of type inference: A mapping from expressions and patterns to types.
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct InferenceResult {
     /// For each method call expr, records the function it resolves to.
     method_resolutions: FxHashMap<ExprId, Function>,
     /// For each field access expr, records the field it resolves to.
     field_resolutions: FxHashMap<ExprId, StructField>,
+    /// For each associated function call expr, records the function it resolves to
+    assoc_fn_resolutions: FxHashMap<ExprId, Function>,
```

Comment:
this could be a `FxHashMap<ExprOrPatId, ImplItem>`, no need to be so specific to functions